### DISCUSSION TITLE
Future of picamera2 WebUI (don't worry its good)

### DISCUSSION THREAD

**Discussion by monkeymademe** [monkeymademe/CamUI#103] [Ideas]
So.. After actually using my webui in the wild for some photograpy (You can view the results here: https://www.flickr.com/photos/raspberryjamberlin/albums/72177720321112546) I started to see some holes in the interface and also some functional issues and possible improvements so I wanted to jot down some notes and get some opinions on the topic.

First to make how WebUI works right now: When the cameras are detected they will start up a live stream using one of the encoders and basically have video available to view on the website when its accessed. This issue is that any still photos you take are not really raw from the sensor but are a still from the video essentially... I mean ok its not that simple but its kinda how it works.

My proposal is to have different modes and how the feeds work depend on what you want to do with it. Actually very much like a DSLR or any handheld camera works right now.

**Streaming Mode:** This is the default mode and is essentially what webui is doing right now. This is the feed open on boot and the URLs for pushing to octoprint, obs or homeassistant (when that works reliably) will be available. You can make setting changes etc as it is right now.

**Camera Mode:** Same camera control settings but without the live feed (optional). If you take a photo you would be using the still capture mode in picamera2 which 'should' improve quality. Live feed would be an option you would have to enable if you wanted to see if/what changes effect your still image. But generally speaking you should not see the video mode unless you want it and if its on and you want to take a picture the mode will need to switch to still and then take the image. This should improve performance also cause the stream is not running all the time so you can max out the resolutions for your still images without having to worry if the pi can do it. This mode should also support full screen controls in a UI so that if you have a touch screen you can hold the camera up like a camera and make adjustments.

**Video Mode:** In theory it should be similar to the streaming mode but the logistics of video recording are unknown to me. For now it would be assumed that the live feed would be available for setting control and then after the feed is stopped and the recording starts. Or both recording and live feed are available at the same time. Also Audio is a very good question if that would be available.

**Timelapse Mode:** This one is tricky... There are 2 ways to do this. Take stills from the feed like the streaming mode or using the still capture mode in picamera2. If there is a improved quality in the photos using Camera Mode then I would consider using both and allow the user to choose which mode... Maybe... Thats also a question for intervals. If using the Live feed you can have much quicker intervals than with the still capture mode so the option to use one or the other would be based on that factor.

**AI Mode:** Also one thats up for debate on how it would work. To start off with it would only be available for the AI camera and likely just the current demos that are available. My hope maybe is to tie into the detection models (which as a user you could update) and you could set actions like still capture based on triggers! Kinda cool.

So after the bugs/tickets are complete for 1.0.15 I will be looking to start changing the structure of the UI for 1.0.16+. The 1st mode to be added would be the AI mode, at least at a basic level just because thats a hot topic at the moment.

Are there any modes I am missing? Anything you would like to be added?

**Comment by monkeymademe** [comment]
Actually 1st hurdle I just want to put out there... Would 2 cameras connected to the Pi mean you can have 2 different modes running and my thought it yes within reason.

Camera mode for example would likely select the 1st detected camera and allow you to select the other camera if its available. But all other modes would be suspended. I mean why have the 2nd camera streaming to nothing when its not used

BUT

Timelapse and streaming modes might be mixed. Lets say you have a print and you want to live stream one view and timelapse the other... why not?

Or Timelapse and AI mode... have the AI camera detect objects and the other making a timelapse so you could detect a bird and flag the time it was there but still record a timelapse using the other camera.

Could have some interesting use cases

**Reply by Nama** [reply]
Maybe structure it so, that you can set a mode per camera.

**Comment by Nama** [comment]
tl;dr

Features I'd like
- Scheduling (I started looking into it myself)
- motion sensor ([gpiozero](https://gpiozero.readthedocs.io/en/latest/recipes.html#motion-sensor), need capture video before I could look into it)
- trigger script after being done with capture
- control motor (ULN2003APG +  28BYJ-48)


### Full text
I recently bumped into this project, while looking for [this](https://elinux.org/RPi-Cam-Web-Interface) XD
I previously stopped using that project because I wanted to run arch and its heavily dependent on raspbian. (Now I'm using dietPi, so thats not relevant anymore.)

I like the idea of not having the stream active all the time. You probably will be on the stream more often on the first few days and then like never check it for weeks. Don't having any processing for nothing sounds good!

Currently I'm using picamera2 for scheduled stills, which are fed into a script which sends me the stills.  
So I'd like to having a scheduling feature (I started looking into it myself).

The other Pi has a motion sensor and captures 20 seconds videos on trigger. Since you already use gpiozero for buttons, that shouldn't be hard to implement. I can look into this, after the restructure you are talking about. Since recording videos is missing completely now.

The most reason I wanna switch from my scripts to this project is, that I sometimes want to look at the live feed or capture a video instead of stills. And it feels like, its not that easy to have all three at the same time. And since you have the live feed already, I wanted to check if it would be possible to implement recording into it. But, as I assumed, you are already planning that.

I have one stepper motor (ULN2003APG +  28BYJ-48) unused, might as well use it on the Pi with the motion sensor and make the motor move via the WebUI.
[gpiozero](https://gpiozero.readthedocs.io/en/latest/recipes.html#motors) doesn't seem to have features for that tho.

**Reply by TimPGamer** [reply]
That idea with the Motors Even with other purposes is Great i had that also a weile back and U r Right Video would be just awesome. So lets pray it dienst take too Long, but i think we r gonna have that before GTA6. If not, before it comes for PC. ;)

**Comment by monkeymademe** [comment]
Hey sorry I have been offline of late christmas and such...

I'll be looking over all the comments and building a new roadmap for 2025. Some good ideas there and I need to mix them into what I already have. The AI thing took a back seat for complexity so I will look into the usability comments and the camera modes next... More to come

**Comment by phantos** [comment]
The camera mode that is proposed.  That would allow changes to the camera settings while it is streaming under another process as long as the live feed is disabled?  That would be a great addition.

**Comment by monkeymademe** [comment]
Yeah to add the camera mode and to be honest every other mode feels like it needs a total rewrite of all functions and how settings are made... pretty much everything. I have a Repo called https://github.com/monkeymademe/picamera2-WebUI-beta which I will update to this branch and basically butcher the whole thing to get modes working. 

Basically pull the focus away from streaming the feed.

**Comment by Bluesguy1984** [comment]
I would like to be able to stream to my motioneye server.  At this time, motioneye can register the camera as an MJPEG streaming device.  However, you can't save the video, so that means if I want to use it as part of my security camera system, I don't have video being recorded.
